draft-ietf-mls-protocol.md

```diff
@@ -2029,13 +2029,59 @@ is the root, then `parent_hash` is set to a zero-length octet string.
 Otherwise `parent_hash` is the Parent Hash of P's parent with P's sibling as the
 co-path child.
 
-Finally, `original_child_resolution` is the array of `HPKEPublicKey` values of the
-nodes in the resolution of S but with the `unmerged_leaves` of P omitted. For
-example, in the ratchet tree depicted in {{ratchet-tree-nodes}} the
-`ParentHashInput` of node Z with co-path child C would contain an empty
-`original_child_resolution` since C's resolution includes only itself but C is also
-an unmerged leaf of Z. Meanwhile, the `ParentHashInput` of node Z with co-path child
-D has an array with one element in it: the HPKE public key of D.
+Finally, `original_sibling_tree_hash` is the original tree hash of S. The
+original tree hash corresponds to the tree hash of S the last time P was
+updated. It can be computed as the tree hash of S in the ratchet tree modified
+the following way:
+
+* reset the leaves in P.unmerged_leaves to blanks
+* remove P.unmerged_leaves from all unmerged_leaves lists
+* truncate the ratchet tree as described in {{remove}}
+
```

Comment:
```suggestion
Note that no recomputation is needed if the tree hash of S is unchanged since
the last time P was updated.  This is the case for computing or processing a
Commit whose UpdatePath traverses P, since the Commit itself resets P.  (In
other words, it is only necessary to recompute the original sibling tree hash
when validating group's tree on joining.) More generally, if none of the entries
in `P.unmerged_leaves` is in the subtree under S (and thus no nodes were truncated),
then the original tree hash at S is the tree hash of S in the current tree.

If it is necessary to recompute the original tree hash of a node, the efficiency
of recomputation can be improved by caching intermediate tree hashes, to avoid
recomputing over the subtree when the subtree is included in multiple parent
hashes.  A subtree hash can be reused as long as the intersection of the
parent's unmerged leaves with the subtree is the same as in the earlier
computation.
```